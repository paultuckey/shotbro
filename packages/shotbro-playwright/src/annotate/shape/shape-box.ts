import type {BoxShape} from "./shape-types";
import type {ShotBroBox} from "../annotate-types";
import {ulid} from "../../util/ulid";

const defaultProps: BoxShape = {
    thickness: 4,
    //padding: 4,
    color: 'deeppink',
    translate: 0,
    translateX: 0,
    translateY: 0,
}

export async function renderBox(elPos: ShotBroBox, rawShape: BoxShape): Promise<string> {
    const shape: BoxShape = Object.assign({}, defaultProps, rawShape);
    const scope = 'box-' + ulid();
    const id = rawShape.id || scope;
    const thickness = shape.thickness || 0;
    const translateX = shape.translateX ?? shape.translate ?? 0
    const translateY = shape.translateY ?? shape.translate ?? 0
    return `
        <style>
            .box.${scope} {
                position: fixed;
                top: ${elPos.y + translateY - (thickness / 2)}px;
                left: ${elPos.x + translateX - (thickness / 2)}px;
                width: ${(thickness * 2) + elPos.w}px;
                height: ${(thickness * 2) + elPos.h}px;
                
                border-width: ${(shape.thickness || 0)}px;
                border-style: solid;
                border-color: ${shape.color};
                border-radius: ${thickness}px;
                box-shadow: ${thickness / 2}px ${thickness / 2}px 2px 0px rgba(0,0,0,0.13);
            }
        </style>
        <div id="${id}" class="box ${scope}"></div>`;
}
