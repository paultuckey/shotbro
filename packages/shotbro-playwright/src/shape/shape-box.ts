import type {BoxShape, ShotBroShape} from "./shape-types";
import type {ShotBroBox} from "../annotate/annotate-types";
import {BLEED} from "../overlay/overlay-html-generator";
import {n2dp} from "./shape-utils";

const defaultProps: BoxShape = {
    thickness: 4,
    color: 'deeppink',
    translate: 0,
    translateX: 0,
    translateY: 0,
    rotate: 0,
}

export async function renderBox(scope: string, elPos: ShotBroBox, rawShape: ShotBroShape): Promise<string> {
    const shape: BoxShape = Object.assign({}, defaultProps, rawShape.box);
    const id = shape.id || scope;
    const thickness = shape.thickness || 0;
    const translateX = shape.translateX ?? shape.translate ?? 0
    const translateY = shape.translateY ?? shape.translate ?? 0
    return `
        <style>
            .${scope}.box {
                position: absolute;
                top: ${n2dp(elPos.y - (thickness / 2) + translateY + BLEED)}px;
                left: ${n2dp(elPos.x - (thickness / 2) + translateX + BLEED)}px;
                width: ${elPos.w + thickness}px;
                height: ${elPos.h + thickness}px;
                transform: rotate(${shape.rotate}deg);
                
                border-width: ${(shape.thickness || 0)}px;
                border-style: solid;
                border-color: ${shape.color};
                border-radius: ${thickness}px;
            }
        </style>
        <div id="${id}" class="${scope} box"></div>`;
}
