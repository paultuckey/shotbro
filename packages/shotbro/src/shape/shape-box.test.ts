import {renderBox} from "./shape-box";
import type {ShotBroBox, BoxShape} from "./shape-types";

describe('Shape Box', () => {

  test('Simple test', async () => {
    const shape: BoxShape = {
      //thickness: 'Hello World'
    }
    const elPos: ShotBroBox = {x: 1, y: 2, w: 40, h: 50}
    const html = await renderBox(elPos, shape);
    expect(html).toMatchSnapshot('box-simple');
  })
});