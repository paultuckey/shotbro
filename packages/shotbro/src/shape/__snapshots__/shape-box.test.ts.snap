// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shape Box Simple test: box-simple 1`] = `"<div style="position:fixed;top:0px;left:-1px;width:48px;height:58px;border-width:4px;border-style:solid;border-color:deeppink;border-radius:4px;box-shadow:2px 2px 2px 0px rgba(0,0,0,0.13);"></div>"`;
